### PR TITLE
Santec extras

### PR DESCRIPTION
**Pull request by ichristen**
Subsumes #10 .

### CODE REVIEW

File: slmsuite/hardware/slms/santec.py [hunk 1]
```diff
@@ -90,85 +100,157 @@ def __init__(self, slm_number=1, display_number=2, verbose=True, **kwargs):
...
+            Santec._parse_status(
+                slm_funcs.SLM_Ctrl_ReadWL(self.slm_number, wav_current_nm, phase_current)
+            )
+
+            wav_desired_nm = int(1e3 * wav_design_um)
+            phase_desired = int(np.floor(max_phase * 100 / np.pi))
+
+            attempt = 1
+
+            while wav_current_nm.value != wav_desired_nm and attempt < 5:
```
Comment (tpr0p) [inline slmsuite/hardware/slms/santec.py:143]: Why multiple attempts?

File: slmsuite/hardware/slms/santec.py [hunk 2]
```diff
@@ -258,15 +345,115 @@ def close(self):
...
+        for drive_error_bit in slm_funcs.SLM_DRIVEBOARD_ERROR.keys():
+            if drive_error.value & drive_error_bit:
+                errors.append(slm_funcs.SLM_DRIVEBOARD_ERROR[drive_error_bit])
+
+        for option_error_bit in slm_funcs.SLM_OPTIONBOARD_ERROR.keys():
+            if option_error.value & option_error_bit:
+                errors.append(slm_funcs.SLM_OPTIONBOARD_ERROR[option_error_bit])
+
+        if raise_error and len(errors) > 0:
+            raise RuntimeError("Santec error: " + ", ".join(["'" + err + "'" for err in errors]))
```
Comment (tpr0p) [inline slmsuite/hardware/slms/santec.py:401]: Not worth redoing, but something to think about for future string building is that `"'" + err + "'"` creates 2 different strings (joins the first two then joins the combined first two with the third) and is somewhat less powerful than the format string construction, which lets you cast types and specify how many, e.g. decimals to display
[https://docs.python.org/3/library/string.html#formatstrings](https://docs.python.org/3/library/string.html#formatstrings)
So I would've done
"'{}'".format(err)
but they're basically the same and it doesn't really matter

File: slmsuite/hardware/slms/slm.py [hunk 3]
```diff
@@ -202,35 +220,61 @@ def write(
...
+               This is error-checked such that bits with greater significance than the
+               bitdepth of the SLM are zero (e.g. the final 6 bits of 16 bit data for a
+               10-bit SLM). Integer data with type different from :attr:`display` leads
+               to an AssertionError.
+
+            Usually, an **exact** stored copy of the data passed by the user under
+            ``phase`` is stored in the attribute :attr:`phase`.
+            However, in cases where :attr:`phase_scaling` not one, this
+            copy is modified to include how the data was wrapped. If the data was
+            cropped, then the cropped data is stored, etc. If integer data was passed,
```
Comment (tpr0p) [inline slmsuite/hardware/slms/slm.py:276]: I feel like the correct thing to do is to calculate phase information (float) for the passed integer data.

**Review by tpr0p — CHANGES_REQUESTED** [review]
1) In santec.py, add comment about why multiple attempts are needed.
2) In slm.py, integer data should be converted to float data, though I could be convinced that this is unnecessary or a bad idea. I both want the user to have information about the phase and not think that the old phase is what's currently on the display.

File: slmsuite/hardware/slms/santec.py (hunk 2 above)
Comment (ichristen) [reply slmsuite/hardware/slms/santec.py:401]: This code would have square brackets, but I guess it's easier to strip those.

File: slmsuite/hardware/slms/slm.py (hunk 3 above)
Comment (ichristen) [reply slmsuite/hardware/slms/slm.py]: Fixed in cdb1a12.

File: slmsuite/hardware/slms/santec.py (hunk 1 above)
Comment (ichristen) [reply slmsuite/hardware/slms/santec.py:145]: Fixed in cdb1a12.

File: slmsuite/holography/analysis.py [hunk 4]
```diff
@@ -230,8 +230,8 @@ def image_moment(images, moment=(1, 0), centers=(0, 0), normalize=True, nansum=F
 
     Warning
     ~~~~~~~
-    This function does not check if the images in ``images`` are non-negative, or correct
-    for this. Negative values may produce unusual results.
+    This function does not check if the passed ``images`` are not negative, or correct
```
Comment (cpanuski) [inline slmsuite/holography/analysis.py:233]:
```suggestion
    This function does not check (or correct for) for passed ``images`` with negative values.
```

**Review by cpanuski — CHANGES_REQUESTED** [review]

**Pull request merged**